### PR TITLE
Adding a few more struct scenarions and a ResultValidation()

### PR DESCRIPTION
**Pull request by Srinivasuluch**
Adding more struct scenarions and a ResultValidation()

### CODE REVIEW

File: tests/src/kernel/hipLaunchParm.cpp [hunk 1]
```diff
@@ -23,75 +23,768 @@ THE SOFTWARE.
...
+    hipLaunchKernelStruct_h7.si = 1;
+    #if ENABLE_ALIGNMENT_TEST  // This is broken on small bar
+    hipLaunchKernelGGL(HIP_KERNEL_NAME(hipLaunchKernelStructFunc7),
+                    dim3(BLOCK_DIM_SIZE),
+                    dim3(1), 0, 0, hipLaunchKernelStruct_h7,
+                    result_d);
+    ResultValidation();
+    #endif
+
+    // Test: Passing struct with packed aligned to 6Bytes
```
Comment (yxsamliu) [inline tests/src/kernel/hipLaunchParm.cpp]: aligned to 4 bytes

File: tests/src/kernel/hipLaunchParm.cpp [hunk 2]
```diff
@@ -23,75 +23,768 @@ THE SOFTWARE.
...
+                    bool* result_d8) {
+    int x = blockIdx.x * blockDim.x + threadIdx.x;
+
+    // set the result to true if the condition met
+    // get the address of the xth element, struct[x],
+    // size_t(p)%4 will be 0 if aligned to 4Byte address space
+    int *p = (int*)(&hipLaunchKernelStruct_);
+    result_d8[x] =  ((hipLaunchKernelStruct_.c1 == 'c')
+                      && (hipLaunchKernelStruct_.si == 1)
+                      && ((size_t(p))%4 ==0) );
```
Comment (yxsamliu) [inline tests/src/kernel/hipLaunchParm.cpp]: also need to check sizeof(hipLaunchKernelStruct_)==6

File: tests/src/kernel/hipLaunchParm.cpp [hunk 3]
```diff
@@ -23,75 +23,768 @@ THE SOFTWARE.
...
+    #if ENABLE_ALIGNMENT_TEST  // This is broken on small bar
+    hipLaunchKernelGGL(HIP_KERNEL_NAME(hipLaunchKernelStructFunc7),
+                    dim3(BLOCK_DIM_SIZE),
+                    dim3(1), 0, 0, hipLaunchKernelStruct_h7,
+                    result_d);
+    ResultValidation();
+    #endif
+
+    // Test: Passing struct with packed aligned to 6Bytes
+    hipLaunchKernelStruct_t8 hipLaunchKernelStruct_h8;
```
Comment (yxsamliu) [inline tests/src/kernel/hipLaunchParm.cpp]: can you add another test just to check packed struct size is expected? You can reuse this structure type, just remove the aligned attribute but keep the packed attribute.

Comment (yxsamliu) [reply tests/src/kernel/hipLaunchParm.cpp]: can you also add another test which has struct type argument with an aligned data member?

File: tests/src/kernel/hipLaunchParm.cpp [hunk 4]
```diff
@@ -23,75 +23,768 @@ THE SOFTWARE.
...
+    // allocating memory for char pointer on device
+    HIPCHECK(hipMalloc((void**)&ip_d9, sizeof(uint32_t)));
+    HIPCHECK(hipMemset(ip_d9, 1, sizeof(uint32_t)));
+    // ip_d9 passed as pointer to struct member, struct.ip = &ip_d9
+    const hipLaunchKernelStruct_t9 hipLaunchKernelStruct_h9 = {'c', ip_d9};
+    hipLaunchKernelGGL(HIP_KERNEL_NAME(hipLaunchKernelStructFunc9),
+                    dim3(BLOCK_DIM_SIZE),
+                    dim3(1), 0, 0, hipLaunchKernelStruct_h9,
+                    result_d);
+    #if ENABLE_DECLARE_INITIALIZATION_POINTER
```
Comment (yxsamliu) [inline tests/src/kernel/hipLaunchParm.cpp]: does this test fail with hip-clang? This is unexpected.

File: tests/src/kernel/hipLaunchParm.cpp [hunk 5]
```diff
@@ -23,75 +23,768 @@ THE SOFTWARE.
...
 
+#include <cstdint>
 #include "hip/hip_runtime.h"
 #include "test_common.h"
-#include "hip/hip_runtime_api.h"
-#include <iostream>
 
-__global__ void vAdd(hipLaunchParm lp, float* a) {}
+// Memory alignment is broken
+#define ENABLE_ALIGNMENT_TEST 0
```
Comment (yxsamliu) [inline tests/src/kernel/hipLaunchParm.cpp]: Are the failed tests fail with both hcc and hip-clang?
If not, you can use `#if __HCC__` to differentiate them. Turn off the the test for only hcc or hip-clang.

File: tests/src/kernel/hipLaunchParm.cpp [hunk 6]
```diff
@@ -23,75 +23,768 @@ THE SOFTWARE.
...
+typedef struct hipLaunchKernelStruct14 {
+  int readint;
+  int writeint[BLOCK_DIM_SIZE];  // will write to this from device
+} hipLaunchKernelStruct_t14;
+
+// This test is to verify struct with dynamic memory, new int
+// the heap memory will be accessed from device
+typedef struct hipLaunchKernelStruct15 {
+  char c1;
+  int* heapmem = new int[BLOCK_DIM_SIZE];
```
Comment (yxsamliu) [inline tests/src/kernel/hipLaunchParm.cpp]: device code cannot access host heap memory. This memory has to be allocated by hipMalloc.

File: tests/src/kernel/hipLaunchParm.cpp [hunk 7]
```diff
@@ -23,75 +23,768 @@ THE SOFTWARE.
...
+// This test is to verity user defined STL, simple stack implementation
+typedef struct stackNode {
+    int data;
+    stackNode* nextNode = NULL;
+} stackNode_t;
+typedef struct hipLaunchKernelStruct19 {
+  stackNode_t* stack = NULL;
+  unsigned int size_ = 0;
+  void pushMe(int value) {  // not a device function, setting from host
+    stackNode_t* newNode = new stackNode_t;
```
Comment (yxsamliu) [inline tests/src/kernel/hipLaunchParm.cpp]: device code cannot access host heap memory, so this does not work.
For HCC, there is device function `__hip_hc_malloc` and `__hip_hc_free`, which you may use to implement device function pushMe and popMe. For host code, you may want to use hipMalloc and hipFree.

File: tests/src/kernel/hipLaunchParm.cpp [hunk 8]
```diff
@@ -23,75 +23,768 @@ THE SOFTWARE.
...
+}
+
+// Passing out of order initalized struct, access in-order
+__global__ void hipLaunchKernelStructFunc20(
+                    hipLaunchKernelStruct_t20 hipLaunchKernelStruct_,
+                    bool* result_d20) {
+    int x = blockIdx.x * blockDim.x + threadIdx.x;
+
+    // accessing struct members in order
+    result_d20[x] = (hipLaunchKernelStruct_.name = 'A'
```
Comment (yxsamliu) [inline tests/src/kernel/hipLaunchParm.cpp]: Should = be ==?

File: tests/src/kernel/hipLaunchParm.cpp [hunk 9]
```diff
@@ -23,75 +23,768 @@ THE SOFTWARE.
...
+    // strcmp(hipLaunchKernelStruct_.name, "AMD") -> strcmp is not broken
```
Comment (yxsamliu) [inline tests/src/kernel/hipLaunchParm.cpp]: I am not quite get it. name is a member of char type, why is it compared with a string?

File: tests/src/kernel/hipLaunchParm.cpp [hunk 10]
```diff
@@ -23,75 +23,768 @@ THE SOFTWARE.
...
+    MY_LAUNCH_WITH_PAREN(hipLaunchKernelGGL(vAdd, dim3(1024),
+                         dim3(1), 0, 0, Ad), true, "firstCall");
 #endif
 
+    HIPCHECK(hipHostFree(result_h));
+    HIPCHECK(hipFree(result_d));
+
     passed();
-}
+}
```
Comment (yxsamliu) [inline tests/src/kernel/hipLaunchParm.cpp]: miss newline at end of file

File: tests/src/kernel/hipLaunchParm.cpp (hunk 1 above)
Comment (Srinivasuluch) [reply tests/src/kernel/hipLaunchParm.cpp]: Typo corrected

File: tests/src/kernel/hipLaunchParm.cpp (hunk 2 above)
Comment (Srinivasuluch) [reply tests/src/kernel/hipLaunchParm.cpp]: i guess it's 4bytes not 6, sizeof(hipLaunchKernelStruct_)==4.
Done. Add the size validation condition to the kernel

File: tests/src/kernel/hipLaunchParm.cpp (hunk 3 above)
Comment (Srinivasuluch) [reply tests/src/kernel/hipLaunchParm.cpp]: Done, added two more tests as you suggested, aligned(4) & packed. 
Test names:
hipLaunchKernelStructFunc8A for packed
hipLaunchKernelStructFunc8B for aligned(4)

Results: 
Packed broken
alignment working fine.

Comment (Srinivasuluch) [reply tests/src/kernel/hipLaunchParm.cpp]: To be more specific about aligment member:  with latest changes, alignment is working fine on large bar setup but still it's broken on small bar hence tests disabled

File: tests/src/kernel/hipLaunchParm.cpp (hunk 4 above)
Comment (Srinivasuluch) [reply tests/src/kernel/hipLaunchParm.cpp]: Yes Sam, this is not failing.

File: tests/src/kernel/hipLaunchParm.cpp (hunk 5 above)
Comment (Srinivasuluch) [reply tests/src/kernel/hipLaunchParm.cpp]: Yes, failing on both hcc and hip-clang?

File: tests/src/kernel/hipLaunchParm.cpp (hunk 6 above)
Comment (Srinivasuluch) [reply tests/src/kernel/hipLaunchParm.cpp]: Done

File: tests/src/kernel/hipLaunchParm.cpp (hunk 8 above)
Comment (Srinivasuluch) [reply tests/src/kernel/hipLaunchParm.cpp]: Done

File: tests/src/kernel/hipLaunchParm.cpp (hunk 9 above)
Comment (Srinivasuluch) [reply tests/src/kernel/hipLaunchParm.cpp]: initially i tried with a char* name = "AMD" and called strcmp(). As it was not working i changed to a char instead a pointer

File: tests/src/kernel/hipLaunchParm.cpp (hunk 10 above)
Comment (Srinivasuluch) [reply tests/src/kernel/hipLaunchParm.cpp]: Done

File: tests/src/kernel/hipLaunchParm.cpp (hunk 9 above)
Comment (yxsamliu) [reply tests/src/kernel/hipLaunchParm.cpp]: strcmp is not a device function, therefore cannot be called by a device function.

File: tests/src/kernel/hipLaunchParm.cpp (hunk 7 above)
Comment (Srinivasuluch) [reply tests/src/kernel/hipLaunchParm.cpp]: Done

**Pull request merged**